tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -740,4 +740,180 @@ impl ServerKey {
+                            .iter_mut()
+                            .zip(a.blocks()[first_block_to_add..last_block_to_add + 1].iter())
+                        {
+                            self.key.unchecked_add_assign(ct_left_i, ct_right_i);
+                        }
+                    }
+
+                    // last carry is not interesting
+                    let mut carry_blocks = s.blocks()
+                        [first_block_where_addition_happenned..last_block_where_addition_happenned]
+                        .to_vec();
+
+                    let message_blocks = s.blocks_mut();
+
+                    rayon::join(
+                        || {
+                            message_blocks[first_block_where_addition_happenned
+                                ..last_block_where_addition_happenned + 1]
+                                .par_iter_mut()
+                                .for_each(|block| {
+                                    self.key.message_extract_assign(block);
+                                });
+                        },
+                        || {
+                            carry_blocks.par_iter_mut().for_each(|block| {
+                                self.key.carry_extract_assign(block);
+                            });
+                        },
+                    );
+
+                    let mut carry_ct = RadixCiphertext::from(carry_blocks);
+                    let num_blocks_to_add = s.blocks().len() - carry_ct.blocks.len();
+                    self.extend_radix_with_trivial_zero_blocks_lsb_assign(
+                        &mut carry_ct,
+                        num_blocks_to_add,
+                    );
+                    let carry_ct = T::from(carry_ct.blocks);
+                    (s.clone(), carry_ct)
+                })
+                .collect_into_vec(&mut tmp_out);
+
+            // tmp_out elements are tuple of 2 elements (message, carry)
+            let num_ct_created = tmp_out.len() * 2;
+            // Ciphertexts not treated in this iteration are at the end of ciphertexts vec.
+            // the rotation will make them 'wrap around' and be placed at range index
+            // (num_ct_created..remainder_len + num_ct_created)
+            // We will then fill the indices in range (0..num_ct_created)
+            ciphertexts.rotate_right(remainder_len + num_ct_created);
+
+            // Drain elements out of tmp_out to replace them
+            // at the beginning of the ciphertexts left to add
+            for (i, (m, c)) in tmp_out.drain(..).enumerate() {
+                ciphertexts[i * 2] = m;
+                ciphertexts[(i * 2) + 1] = c;
+            }
+            ciphertexts.truncate(num_ct_created + remainder_len);
+        }
+
+        // Now we will add the last chunk of terms
+        // just as was done above, however we do it
+        // we want to use an addition that leaves
+        // the resulting ciphertext with empty carries
```

Comment:
Here I chose to leave clean carries because depending on the number of additions done in that last part, you could be in a state where the only thing possible is to clean carries

Reply:
Also at the time when this was written, the full propagate was still using  the slow algorithm. 

Also now that I think of it, there if that last part actually sums `num_elements_to_fill_carry` then the non parallel carry propagation will not work :thinking: